ReactWindows/ReactNative.Shared/UIManager/AccessibleAutomationPeer.cs

```diff
@@ -0,0 +1,104 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using ReactNative.UIManager.Events;
+#if WINDOWS_UWP
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Automation.Peers;
+using Windows.UI.Xaml.Automation.Provider;
+#else
+using System.Windows;
+using System.Windows.Automation.Peers;
+using System.Windows.Automation.Provider;
+#endif
+
+namespace ReactNative.UIManager
+{
+    /// <summary>
+    /// Automation peer for a FrameworkElement that implements IAccessible.
+    /// </summary>
+    /// <typeparam name="T">Type of owner of the automation peer.</typeparam>
+    public sealed class AccessibleAutomationPeer<T> : FrameworkElementAutomationPeer, IInvokeProvider
+        where T : FrameworkElement, IAccessible
+    {
+        /// <summary>
+        /// Hides base UIElement Owner to provide stronger-typed T Owner.
+        /// </summary>
+        private new T Owner => (T)base.Owner;
+
+        /// <inheritdoc />
+        public AccessibleAutomationPeer(T owner)
+            : base(owner)
+        {
+        }
+
+        /// <inheritdoc />
+        protected override IList<AutomationPeer> GetChildrenCore()
+        {
+            bool shouldHideChildren =
+                Owner.ImportantForAccessibility == ImportantForAccessibility.Yes
+                || Owner.ImportantForAccessibility == ImportantForAccessibility.NoHideDescendants;
+            return shouldHideChildren ? null : base.GetChildrenCore();
+        }
+
+        /// <inheritdoc />
+        protected override bool IsContentElementCore()
+        {
+            return Owner.ImportantForAccessibility == ImportantForAccessibility.Yes;
+        }
+
+        /// <inheritdoc />
+        protected override bool IsControlElementCore()
+        {
+            return Owner.ImportantForAccessibility == ImportantForAccessibility.Yes;
+        }
+
+        /// <inheritdoc />
+        protected override AutomationControlType GetAutomationControlTypeCore()
+        {
+            if (Owner.AccessibilityTraits?.Contains(AccessibilityTrait.Button) == true)
+            {
+                return AutomationControlType.Button;
+            }
+            return base.GetAutomationControlTypeCore();
+        }
+
+        /// <inheritdoc />
+        protected override string GetNameCore()
+        {
+            string baseName = base.GetNameCore();
+            if (!string.IsNullOrWhiteSpace(baseName)
+                || Owner.ImportantForAccessibility != ImportantForAccessibility.Yes)
```

Comment:
>Owner.ImportantForAccessibility != ImportantForAccessibility.Yes [](start = 19, length = 64)

Should we invoke a ShouldHideDescendents helper here to make the intent clearer?

Reply:
no, because shouldHideChildren includes case of NoHideDescendants, and in this case we do want to return baseName. but I agree intent should be clearer and I will refactor code a little to make logic clear.

---
In reply to: [159566545](https://github.com/reseul/react-native-windows/pull/1#discussion_r159566545) [](ancestors = 159566545)

Reply:
We need to be consistent with iOS/Android. If we think iOS made the wrong decision, we should consider changing iOS.

---
In reply to: [159785578](https://github.com/reseul/react-native-windows/pull/1#discussion_r159785578) [](ancestors = 159785578,159566545)